### DISCUSSION TITLE
Size of binary

### DISCUSSION THREAD

**Discussion by grant-guo** [ClickHouse/ClickHouse#29550] [Q&A] [answered]
I tried to compile the source code of the branch `v21.9.4.35-stable` on `Ubuntu 20.04.3 LTS` 

The size of the binary `clickhouse` is 474M, instead the official binary is 100M smaller

here is how I built

```
mkdir build && cd build
CC=/usr/bin/gcc-10 CXX=/usr/bin/g++-10 cmake -DCMAKE_BUILD_TYPE=Release ..
CC=/usr/bin/gcc-10 CXX=/usr/bin/g++-10 cmake --build .
```

When I used clang-13, always got the following errors:

```
"~/work/git/ClickHouse/contrib/libc-headers/x86_64-linux-gnu/sys/select.h" 128L, 4140C                                                                                                                                                                      1,1           Top
/home/gguo/work/git/ClickHouse/contrib/libc-headers/x86_64-linux-gnu/bits/select.h:35:9: note: expanded from macro '__FD_ZERO'
    int __d0, __d1;                                                           \
        ^
/home/gguo/work/git/ClickHouse/base/common/LineReader.cpp:34:5: error: identifier '__d1' is reserved because it starts with '__' [-Werror,-Wreserved-identifier]
/home/gguo/work/git/ClickHouse/contrib/libc-headers/x86_64-linux-gnu/sys/select.h:88:26: note: expanded from macro 'FD_ZERO'
#define FD_ZERO(fdsetp)         __FD_ZERO (fdsetp)
                                ^
/home/gguo/work/git/ClickHouse/contrib/libc-headers/x86_64-linux-gnu/bits/select.h:35:15: note: expanded from macro '__FD_ZERO'
    int __d0, __d1;                                                           \
              ^
2 errors generated.
```

Any ideas?

Thanks

**Comment by alexey-milovidov** [comment, marked as answer]
It can be due to LTO. 
LTO should be manually enabled with ENABLE_THINLTO CMake option.

**Reply by grant-guo** [reply]
I compiled the latest stable branch `v21.9.4.35-stable` using clang-11 and ENABLE_THINLTO. No errors and the binary was reduced by 40M or so. Although the binary is still 60M larger than the official one, that does help. Thanks

**Reply by alexey-milovidov** [reply]
Also official binaries are using `clang-13`.

**Reply by grant-guo** [reply]
> Also official binaries are using `clang-13`.

how to solve double-underscore error with `clang-13`?

**Reply by alexey-milovidov** [reply]
clang-13 is for newer versions (master).
For old versions, just use the same clang version as was used for release build.

**Comment by grant-guo** [comment]
Mark, `strip` the compiled binary, the size reduces to 340M
